rapid7_insightvm/komand_rapid7_insightvm/actions/asset_search/action.py

```diff
@@ -17,12 +17,44 @@ def __init__(self):
     def run(self, params={}):
 
         resource_helper = ResourceHelper(self.connection.session, self.logger)
-        search_criteria = params.get('searchCriteria')
-        self.logger.info("Performing filtered asset search with criteria %s" % search_criteria)
+        search_criteria = params.get(Input.SEARCHCRITERIA)
+        size = params.get(Input.SIZE, 0)
+        sort = params.get(Input.SORT, '')
+        self.logger.info(f'Performing filtered asset search with criteria {search_criteria}')
         endpoint = endpoints.Asset.search(self.connection.console_url)
 
-        response = resource_helper.paged_resource_request(endpoint=endpoint,
-                                                          method='post',
-                                                          payload=search_criteria)
+        if size == 0:
+            parameters = {'sort': sort}
+            resources = resource_helper.paged_resource_request(endpoint=endpoint,
+                                                               method='post', params=parameters,
+                                                               payload=search_criteria)
+        elif size <= 100:
+            parameters = {'size': size, 'sort': sort}
+            resources = resource_helper.resource_request(endpoint=endpoint,
+                                                         method='post', params=parameters,
+                                                         payload=search_criteria)
+            resources = resources['resources']
+        else:
+            resources = []
+            current_page = 0
+            page_size = 100
+            results_retreved = 0
+            while results_retreved <= size:
+                if results_retreved + page_size > size:
+                    page_size = size - results_retreved
+                self.logger.info(f"Fetching results from page {current_page}")
+                parameters = {'sort': sort, 'page': current_page, 'size': page_size}
+                response = resource_helper.get_resource_page(endpoint=endpoint,
+                                                             method='post', params=parameters,
+                                                             payload=search_criteria)
 
-        return {"assets": response}
+                resources += response.resources  # Grab resources and append to total
+                results_retreved += len(response.resources)
+                if (response.total_pages == 0) or ((response.total_pages - 1) == response.page_num):
+                    self.logger.info("All pages consumed, returning results...")
+                    break  # exit the loop
+                else:
+                    self.logger.info("More pages exist, fetching...")
+                    current_page += 1
```

Comment:
Can you just divide the desired result size by the maximum amount of results that can be returned per page and calculate that in advance instead of having to loop like this? Say they want 100 results, max page size is 20, so just calculate that 5 calls are needed.

Reply:
We still have to deal with remainders  if they want 105 results and the max is 20 we need 5.25 pages